### PR TITLE
Introduce LLVMCPUTile to get rid of CodegenStrategy.

### PR DESCRIPTION
**Pull request by hanhanW**
Is is a walk-based transformation. It calls scf::tileUsingSCFForOp method to tile all TileInterface operations. The functionality is tested upstream. The LLVMCPU integration tests are covered by pipeline_tests.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUTile.cpp [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+  for (auto computeOp : computeOps) {
+    auto op = cast<TilingInterface>(computeOp);
+    if (op.getLoopIteratorTypes().empty()) continue;
+
+    // For now do not tile `tensor.pad` operations. The `tensor.pad`
+    // operations might be those introduced by the padding-based
+    // codegeneration strategy. Those are not meant to be tiled again.
+    // Need a better way for handling this, but this works for now.
+    if (isa<tensor::PadOp>(computeOp)) continue;
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUTile.cpp:89]: @MaheshRavishankar this should cover your changes in pad_fusion branch. And yes, we don't tile pad ops that are created by DoubleTilingPadExpert.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUTile.cpp [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+#define DEBUG_TYPE "iree-llvmcpu-tile"
+
+namespace mlir {
+namespace iree_compiler {
+namespace {
+
+/// Builds a proper tile sizes vector for the op.
+/// scf::tileUsingSCFForOp expects the num of tile sizes = num of loops. This
+/// method returns a proper tile sizes vector for each op during tiling.
+static SmallVector<Value> buildTileSizesForOp(OpBuilder &b, Operation *op,
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUTile.cpp:30]: Is this method copied from somewhere?

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Nice! This replaces single tiling expert right? It would be good to also merge the PR that remvoes it (I assume you already have it) with this PR. It gives an idea of what is being removed and what is it being replaced with.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUTile.cpp (hunk 2 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUTile.cpp:30]: it is copied from LinalgTensorCodegenDriver.cpp The method in the CodegenDriver file is removed in the following commit.

https://github.com/openxla/iree/blob/d136a6d33c51a5688728a166f835af2d5b3bf92b/compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp#L86-L101

**Review by MaheshRavishankar — APPROVED** [review]

**Review by dcaballe — APPROVED** [review]
Awesome!

**Comment by hanhanW** [comment]
I feel that the regression is as same as https://github.com/openxla/iree/issues/12808

I'll land the PR after confirming it, i.e., there are no regressions.

**Comment by dcaballe** [comment]
It could be related to this canonicalization pattern: https://reviews.llvm.org/D147099

I didn't see i1 transposes while profiling the models but now it looks like some canonicalizations are not happening in the same order as before and that is exposing some issues. For example, it looks like https://reviews.llvm.org/D147099 is now needed because `vector.create_mask` is turned into `vector.constant_mask` before the transposed is removed.

**Comment by hanhanW** [comment]
most of regressions are gone after re-triggering the benchmarks. Looks like it's a false alarm

**Pull request merged**